source/common/router/header_formatter.h

```diff
@@ -98,5 +100,74 @@ class CompoundHeaderFormatter : public HeaderFormatter {
   const bool append_;
 };
 
+/**
+ * HttpHeaderFormatter is used by HTTP headers manipulators.
+ **/
+class HttpHeaderFormatter {
+public:
+  virtual ~HttpHeaderFormatter() = default;
+
+  virtual const std::string format(const Http::RequestHeaderMap& request_headers,
+                                   const Http::ResponseHeaderMap& response_headers,
+                                   const Envoy::StreamInfo::StreamInfo& stream_info) const PURE;
+
+  /**
+   * @return bool indicating whether the formatted header should be appended to the existing
+   *              headers or replace any existing values for the header
+   */
+  virtual bool append() const PURE;
+};
+
+using HttpHeaderFormatterPtr = std::unique_ptr<HttpHeaderFormatter>;
+
+/**
+ * Implementation of HttpHeaderFormatter.
+ * Actual formatting is done via substitution formatters.
+ */
+class HttpHeaderFormatterImpl : public HttpHeaderFormatter {
+public:
+  HttpHeaderFormatterImpl(Formatter::FormatterPtr&& formatter, bool append)
+      : formatter_(std::move(formatter)), append_(append) {}
+
+  // HttpHeaderFormatter::format
+  // Trailers are not available when HTTP headers are manipulated.
+  const std::string format(const Http::RequestHeaderMap& request_headers,
+                           const Http::ResponseHeaderMap& response_headers,
+                           const Envoy::StreamInfo::StreamInfo& stream_info) const override {
+    std::string buf;
+    buf = formatter_->format(request_headers, response_headers,
+                             *Http::StaticEmptyHeaders::get().response_trailers, stream_info, "");
```

Comment:
if trailers are optional, let's make the OptRef as well?

Reply:
@alyssawilk That is a good idea (to use `optref`), but I think it would be better to do it in a separate PR (maybe after runtime guard is deprecated). `format` method is implemented in around 50 access logs formatters and I would have to change them all. It will add probably hundreds of lines to review to this already large PR. I can open an issue to track that activity.